cache/cache_reservation_manager.cc

```diff
@@ -0,0 +1,148 @@
+    // (2) Delayed releasing previously inserted dummy entries can save such
+    // expensive dummy entry insertion on memory increase in the near future,
+    // which is likely to happen when the memory usage is greater than or equal
+    // to 3/4 of what we reserve
+    if (delayed_decrease_ && new_mem_used >= cur_cache_allocated_size / 4 * 3) {
+      return Status::OK();
+    } else {
+      Status s = DecreaseCacheReservation(new_mem_used);
+      return s;
+    }
+  }
+}
+
+// Explicitly instantiate templates for "CacheEntryRole" values we use.
+// This makes it possible to keep the template definitions in the .cc file.
+template Status CacheReservationManager::UpdateCacheReservation<
+    CacheEntryRole::kWriteBuffer>(std::size_t new_mem_used);
+// For cache reservation manager unit tests
+template Status CacheReservationManager::UpdateCacheReservation<
+    CacheEntryRole::kMisc>(std::size_t new_mem_used);
+
+template <CacheEntryRole R>
+Status CacheReservationManager::IncreaseCacheReservation(
+    std::size_t new_mem_used) {
+  Status return_status = Status::OK();
+  while (new_mem_used > cache_allocated_size_.load(std::memory_order_relaxed)) {
+    Cache::Handle* handle = nullptr;
+    Status s = cache_->Insert(GetNextCacheKey(), nullptr, kSizeDummyEntry,
+                              GetNoopDeleterForRole<R>(), &handle);
+
+    // We do not return immeditately when encountering dummy entry insertion
+    // error to keep cache_allocated_size_ easy to reasonate. In other words, we
+    // still account those dummy entries that fail insertion in
+    // cache_allocation_size_. As a consequence, there will exist null dummy
+    // handles in dummy_handles_ if insertion error occurs.
+
+    // Ideallly we should prevent this allocation from happening if
+    // this insertion fails. However, the callers to this code path
+    // might not be able to handle failures properly. [TODO] We'll need to
+    // improve it in the future and figure out what to do on error
+    s.PermitUncheckedError();
+
+    // We do not overwrite any non-OK status from dummy entry insertion.
+    // This is to prevent possible subsequent successful dummy entry insertion
+    // (due to newly available cache space resulting from manipulation in other
+    // threads) from overriding the non-OK status
+    if (return_status == Status::OK()) {
+      return_status = s;
+    }
+
+    dummy_handles_.push_back(handle);
+    cache_allocated_size_ += kSizeDummyEntry;
+  }
+  return return_status;
+}
+
+Status CacheReservationManager::DecreaseCacheReservation(
+    std::size_t new_mem_used) {
+  Status return_status = Status::OK();
+  while (new_mem_used < cache_allocated_size_.load(std::memory_order_relaxed)) {
+    assert(!dummy_handles_.empty());
+    auto* handle = dummy_handles_.back();
+    // If insert failed, handle is null so we should not release.
+    if (handle != nullptr) {
+      cache_->Release(handle, true);
```

Comment:
[Note]
I couldn't find an easy way to see whether an error occurred in Release() so it always returns Status::OK() now. 

I noticed that in the document [here](https://github.com/facebook/rocksdb/blob/master/include/rocksdb/cache.h#L413-L416) , it indicated that we should call "deleter"explicitly when there is an error in releasing. However, I honestly don't know how to know whether there is an error ....